pkg/serverless/trace/cold_start_span_creator.go

```diff
@@ -35,45 +33,90 @@ const (
 var functionName = os.Getenv(functionNameEnvVar)
 
 type ColdStartSpanCreator struct {
-	executionContext *executioncontext.ExecutionContext
-	traceAgent       *ServerlessTraceAgent
-	createSpan       *sync.Once
+	TraceAgent            *ServerlessTraceAgent
+	CreateSpan            *sync.Once
+	LambdaSpanChan        chan *pb.Span
+	InitDurationChan      chan float64
+	SyncSpanDurationMutex *sync.Mutex
+	ColdStartSpanId       uint64
+	lambdaSpan            *pb.Span
+	initDuration          float64
+	StopChan              chan struct{}
 }
 
-func (c *ColdStartSpanCreator) create(lambdaSpan *pb.Span) {
-	// Prevent infinite loop from SpanModifier call
-	if lambdaSpan.Name == "aws.lambda.cold_start" {
+func (c *ColdStartSpanCreator) Run() {
+	go func() {
+		for {
+			select {
+			case span := <-c.LambdaSpanChan:
+				if span.Name == spanName {
+					continue
+				}
+				c.lambdaSpan = span
+				c.createIfReady()
+
+			case initDuration := <-c.InitDurationChan:
+				c.initDuration = initDuration
+				c.createIfReady()
+
+			case <-c.StopChan:
+				log.Debugf("[ColdStartCreator] - shutting down")
+				return
+			default:
```

Comment:
I think it'd be okay to remove the `default` as it could actually save some cpu. The select will then just block until a thing is received on one of the three channels, which is kinda what we want it to do anyway.